packages/clarity-js/src/layout/extract.ts

```diff
@@ -0,0 +1,64 @@
+import { Extract, Metric, Region, RegionFilter } from "@clarity-types/core";
+import { Constant } from "@clarity-types/data";
+import * as metric from "@src/data/metric";
+import * as region from "@src/layout/region";
+
+const formatRegex = /1/g;
+const digitsRegex = /[^0-9\.]/g;
+const digitsWithCommaRegex = /[^0-9\.,]/g;
+
+export function regions(root: ParentNode, value: Region[]): void {
+    for (let v of value) {
+        let valid = true;
+        let filter = v.length === 4 ? v[2] : null;
+        switch (filter) {
+            case RegionFilter.Url: valid = !!top.location.href.match(new RegExp(v[3])); break;
+            case RegionFilter.Javascript: valid = !!evaluate(v[3]); break;
+        }
+        if (valid) { root.querySelectorAll(v[1]).forEach(e => region.observe(e, v[0].toString())); }
+    }
+}
+
+export function metrics(root: ParentNode, value: Metric[]): void {
+    for (let v of value) {
+        let source = v.length >= 3 ? v[1] : null;
+        let scale = v.length === 4 ? v[3] : 1;
+        switch (source) {
+            case Extract.Text:  root.querySelectorAll(v[2]).forEach(e => { metric.max(v[0], num((e as HTMLElement).innerText, scale)); }); break;
+            case Extract.Attribute: root.querySelectorAll(`[${v[2]}]`).forEach(e => { metric.max(v[0], num(e.getAttribute(v[2]), scale, false)); }); break;
+            case Extract.Javascript: metric.max(v[0], evaluate(v[2], Constant.Number) as number); break;
+        }
+    }
+}
+
+function evaluate(variable: string, type: string = null, base: Object = window): any {
```

Comment:
>evaluate [](start = 9, length = 8)

can you add comments to explain the purpose of this

Reply:
done, thanks.

---
In reply to: [643631761](https://github.com/microsoft/clarity/pull/137#discussion_r643631761) [](ancestors = 643631761)